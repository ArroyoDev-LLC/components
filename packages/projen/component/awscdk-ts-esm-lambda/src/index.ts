export * from './awscdk-ts-esm-lambda'
