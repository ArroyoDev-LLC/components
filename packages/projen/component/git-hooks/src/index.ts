export * from './git-hooks'
export * from './lint-staged'
