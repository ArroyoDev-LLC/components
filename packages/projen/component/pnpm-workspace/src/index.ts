export * from './pnpm-workspace.ts'
