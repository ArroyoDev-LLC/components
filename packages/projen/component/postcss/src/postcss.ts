import {
	TypeScriptSourceConfig,
	type TypeScriptSourceConfigPlugin,
} from '@arroyodev-llc/projen.component.typescript-source-file'
import { findComponent } from '@arroyodev-llc/utils.projen'
import type { ObjectLiteralMergeSchema } from '@arroyodev-llc/utils.ts-ast'
import type { Config } from 'postcss-load-config'
import { Component, type Project, type typescript } from 'projen'

export interface PostCSSOptions {
	/**
	 * Config file path.
	 */
	readonly filePath?: string
	/**
	 * Postcss Config.
	 */
	readonly config?: Config
}

export class PostCSS extends Component {
	public static of(project: Project): PostCSS | undefined {
		return findComponent(project, PostCSS)
	}

	readonly file: TypeScriptSourceConfig<Config>

	constructor(project: typescript.TypeScriptProject, options?: PostCSSOptions) {
		super(project)

		this.file = new TypeScriptSourceConfig(
			project,
			options?.filePath ?? 'postcss.config.ts',
			{
				marker: true,
				pluginsProperty: 'plugins',
				recreate: true,
				source: `export default {}`,
			}
		)
	}

	/**
	 * Add postcss plugin.
	 * @param plugin Plugin spec.
	 */
	addPlugin(plugin: TypeScriptSourceConfigPlugin): this {
		const spec = {
			...(!plugin.options
				? {
						spec: (writer) => writer.write(plugin.name),
				  }
				: {}),
			...plugin,
		} satisfies TypeScriptSourceConfigPlugin
		this.file.addPlugin(spec)
		return this
	}

	/**
	 * Merge postcss config.
	 * @param config Config to merge.
	 */
	addConfig(config: ObjectLiteralMergeSchema<Config>) {
		this.file.addConfig(config)
		return this
	}
}
