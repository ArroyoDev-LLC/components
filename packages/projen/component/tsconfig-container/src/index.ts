export * from './tsconfig-container.ts'
