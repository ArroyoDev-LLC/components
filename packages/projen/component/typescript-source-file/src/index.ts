export * from './typescript-source-file'
export * from './typescript-source-config'
export * from './type-script-source-file-cache'
