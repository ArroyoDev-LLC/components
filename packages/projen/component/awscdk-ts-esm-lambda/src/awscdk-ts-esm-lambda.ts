import { existsSync } from 'node:fs'
import { TypeScriptSourceFile } from '@arroyodev-llc/projen.component.typescript-source-file'
import { findComponent, isComponent } from '@arroyodev-llc/utils.projen'
import {
	awscdk,
	Component,
	type FileBase,
	type Project,
	TextFile,
	typescript,
	javascript,
} from 'projen'

export interface AwsCdkTsEsmLambdaOptions {
	/**
	 * Project relative path to lambda function construct.
	 */
	readonly filePath: string
}

/**
 * AWS CDK TypeScript Lambda function with ESM support.
 */
export class AwsCdkTsEsmLambda extends Component {
	public static of(project: Project): AwsCdkTsEsmLambda | undefined {
		return findComponent(project, AwsCdkTsEsmLambda)
	}

	readonly lambdaFile: FileBase
	readonly tsFile: TypeScriptSourceFile

	constructor(project: Project, options: AwsCdkTsEsmLambdaOptions) {
		super(project)

		this.lambdaFile = project.tryFindFile(options.filePath) as FileBase

		const lambdaComponents = this.project.components.filter((comp) =>
			isComponent(awscdk.LambdaFunction, comp),
		)
		this.node.addDependency(...lambdaComponents.map((comp) => comp.node))

		const sourceCode = this.project.components.find(
			(comp) =>
				isComponent(TextFile, comp) && comp.path === this.lambdaFile.path,
		) as TextFile
		// @ts-expect-error would required replacing component or other post synth hackarounds otherwise since we need to replace the file before it gets written
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access
		const source: string = sourceCode.lines.join('\n')
		this.project.tryRemoveFile(this.lambdaFile.path)

		if (!(this.project instanceof typescript.TypeScriptProject)) {
			throw new Error('AwsCdkTsEsmLambda must be used in a TypeScriptProject')
		}

		this.tsFile = new TypeScriptSourceFile(this.project, this.lambdaFile.path, {
			marker: false,
			recreate: true,
			source,
			format: false,
		})
		this.applyTransforms()
		javascript.Eslint.of(project)?.addIgnorePattern?.(this.tsFile.path)
	}

	/**
	 * Apply transforms to the lambda function source file.
	 * @protected
	 */
	protected applyTransforms() {
		this.tsFile.addImport({
			namedImports: ['fileURLToPath'],
			moduleSpecifier: 'node:url',
		})
		this.tsFile.addTransformer((sourceFile) => {
			const functionClass = sourceFile.getClasses()[0]
			if (!functionClass)
				throw new Error('Could not find class in entrypoint file')
			const functionCtor = functionClass.getConstructors()[0]
			if (!functionCtor)
				throw new Error('Could not find class constructor in entrypoint file')
			functionCtor.insertStatements(0, [
				'const __filename = fileURLToPath(import.meta.url);',
				'const __dirname = path.dirname(__filename);',
			])
			sourceFile.formatText()
		})
	}
}

/**
 * Automatically discover and create AWS CDK TypeScript Lambda functions with ESM support.
 */
export class AwsCdkTsEsmLambdaAutoDiscover extends Component {
	constructor(project: Project) {
		super(project)

		const lambdaFunctionFiles = project.files
			.filter((file) => file.path.endsWith('-function.ts'))
			.filter((file) =>
				existsSync(file.absolutePath.replace('-function.ts', '.lambda.ts')),
			)

		for (const lambdaFunctionFile of lambdaFunctionFiles) {
			new AwsCdkTsEsmLambda(project, { filePath: lambdaFunctionFile.path })
		}
	}
}
