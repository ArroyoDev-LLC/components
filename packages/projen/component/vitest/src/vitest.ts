import path from 'node:path'
import { TypeScriptSourceConfig } from '@arroyodev-llc/projen.component.typescript-source-file'
import { Vite } from '@arroyodev-llc/projen.component.vite'
import { cwdRelativePath } from '@arroyodev-llc/utils.projen'
import { Component, JsonFile, type Project, type Task } from 'projen'
import { type TypeScriptProject } from 'projen/lib/typescript'
import { deepMerge } from 'projen/lib/util'
import { type UserWorkspaceConfig } from 'vitest/config'

export enum VitestConfigType {
	WORKSPACE = 'Workspace',
	PROJECT = 'Project',
	CONFIG = 'Config',
}

export interface VitestOptions {
	/**
	 * Vitest config path.
	 */
	configFilePath?: string
	/**
	 * Configuration type.
	 */
	configType?: VitestConfigType
	/**
	 * Vitest settings.
	 */
	settings?: UserWorkspaceConfig
}

export class Vitest extends Component {
	public static of(project: Project): Vitest | undefined {
		const isVitest = (c: Component): c is Vitest => c instanceof Vitest
		return project.components.find(isVitest)
	}

	readonly configType: VitestConfigType
	readonly configFile: TypeScriptSourceConfig<UserWorkspaceConfig>
	readonly options: Required<VitestOptions>
	readonly vite?: Vite
	readonly testWatchTask: Task
	readonly #workspaceProjects: Map<string, Vitest> = new Map<string, Vitest>()

	constructor(
		public readonly project: TypeScriptProject,
		options: VitestOptions = {
			configFilePath: 'vitest.config.ts',
			configType: VitestConfigType.PROJECT,
			settings: {},
		}
	) {
		super(project)
		this.vite = Vite.of(this.project)
		this.options = deepMerge([
			{
				settings: {
					test: {
						name: project.name,
						include: [`${project.testdir}/\*\*/\*.spec.ts`],
					},
				},
			} as VitestOptions,
			options,
		]) as Required<VitestOptions>
		this.project.addDevDeps('vitest')
		this.configType = this.options.configType ?? VitestConfigType.PROJECT
		const configFilePath = this.options.configFilePath ?? 'vitest.config.ts'

		if (this.configType !== VitestConfigType.WORKSPACE) {
			this.project.testTask.exec('vitest', {
				args: ['--run'],
				receiveArgs: true,
			})
		}
		this.testWatchTask =
			this.project.tasks.tryFind('test:watch') ??
			this.project.addTask('test:watch', {
				receiveArgs: true,
				exec: 'vitest',
				description: 'Run tests on changes.',
			})

		this.project.tsconfigDev?.addInclude?.(configFilePath)
		this.project?.eslint?.addOverride?.({
			files: [
				configFilePath,
				...(this.options.settings?.test?.include ?? [this.project.testdir]),
			],
			rules: {
				'import/no-extraneous-dependencies': ['off'],
			},
		})

		this.configFile =
			TypeScriptSourceConfig.withCallExpressionConfig<UserWorkspaceConfig>(
				this.project,
				configFilePath,
				{
					source: `export default ${this.defineName}({})`,
					recreate: true,
					marker: true,
					config: this.options.settings,
				}
			)

		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: [this.defineName],
		})

		if (this.configType === VitestConfigType.PROJECT) {
			this.tryFindWorkspace()?.addProjectConfig?.(this)
		}

		if (this.configType === VitestConfigType.WORKSPACE) {
			new JsonFile(this.project, 'vitest.workspace.json', {
				readonly: true,
				allowComments: false,
				obj: () => Array.from(this.#workspaceProjects.keys()),
			})
		}
	}

	/**
	 * Get the define function name.
	 * @protected
	 */
	protected get defineName(): string {
		return this.configType === VitestConfigType.PROJECT
			? 'defineProject'
			: 'defineConfig'
	}

	/**
	 * Merge vite config if applicable.
	 * @protected
	 */
	protected mergeVite(): void {
		if (!this.vite) return
		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: ['mergeConfig'],
		})
		this.configFile.addImport({
			moduleSpecifier: cwdRelativePath(
				this.project.outdir,
				this.vite.file.absolutePath.split('.').slice(0, -1).join('.')
			),
			defaultImport: 'viteConfig',
		})
		this.configFile.addConfigTransform((cfg, src) => {
			src.addExportAssignment({
				expression: (writer) =>
					writer.write(
						`mergeConfig(viteConfig, ${this.defineName}(${cfg.getFullText()}))`
					),
				isExportEquals: false,
			})
			src.removeDefaultExport()
		})
	}

	/**
	 * Try to find the workspace {@link Vitest} component.
	 */
	tryFindWorkspace(): Vitest | undefined {
		if (this.configType === VitestConfigType.WORKSPACE) return this
		if (this.project.parent) {
			return Vitest.of(this.project.parent)?.tryFindWorkspace?.()
		}
		return
	}

	/**
	 * Add {@link Vitest} component to this workspace.
	 * @param component Target component.
	 */
	addProjectConfig(component: Vitest): this {
		if (this.configType !== VitestConfigType.WORKSPACE)
			throw new Error('No vitest workspace found. Cannot add project config!')
		const relPath = path.format({
			dir: '.',
			base: path.relative(
				path.dirname(this.configFile.absolutePath),
				component.configFile.absolutePath
			),
		})
		this.#workspaceProjects.set(relPath, component)
		return this
	}

	/**
	 * Add config to merge.
	 * @param config Config part.
	 */
	addConfig(config: this['configFile']['__mergeSchema']): this {
		this.configFile.addConfig(config)
		return this
	}

	/**
	 * Add plugin to merge.
	 * @param plugin Plugin part.
	 */
	addPlugin(plugin: this['configFile']['__pluginSchema']) {
		this.configFile.addPlugin(plugin)
		return this
	}

	/**
	 * @inheritDoc
	 */
	preSynthesize() {
		super.preSynthesize()
		this.mergeVite()
	}
}
