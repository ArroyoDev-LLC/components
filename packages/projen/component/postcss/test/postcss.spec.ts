import os from 'node:os'
import path from 'node:path'
import { typescript } from 'projen'
import { Testing } from 'projen/lib/testing'
import { beforeEach, expect, test } from 'vitest'
import { PostCSS } from '../src'

interface TestContext {
	project: typescript.TypeScriptProject
}

beforeEach<TestContext>((ctx) => {
	ctx.project = new typescript.TypeScriptProject({
		name: ctx.meta.id,
		outdir: path.join(os.tmpdir(), ctx.meta.id),
		defaultReleaseBranch: 'main',
	})
})

test<TestContext>('renders as expected', async (ctx) => {
	new PostCSS(ctx.project)
	const synth = Testing.synth(ctx.project)
	expect(synth['postcss.config.ts']).toBeDefined()
	expect(synth['postcss.config.ts']).toMatchSnapshot()
})

test<TestContext>('renders as expected with merged config', async (ctx) => {
	const postcss = new PostCSS(ctx.project)
	postcss.addConfig({
		to: 'something',
	})
	postcss.addConfig({
		from: (writer) => writer.write('resolveFrom()'),
	})
	const synth = Testing.synth(ctx.project)
	expect(synth['postcss.config.ts']).toBeDefined()
	expect(synth['postcss.config.ts']).toMatchSnapshot()
})

test<TestContext>('renders with plugins as expected', async (ctx) => {
	const postcss = new PostCSS(ctx.project)
	postcss
		.addPlugin({
			name: 'tailwindcss',
			moduleImport: {
				moduleSpecifier: 'tailwindcss',
				defaultImport: 'tailwindcss',
			},
		})
		.addPlugin({
			name: 'postcssPxToRem',
			moduleImport: {
				moduleSpecifier: 'postcss-pxtorem',
				defaultImport: 'postcssPxToRem',
			},
			options: {
				propList: ['font'],
			},
		})
	const synth = Testing.synth(ctx.project)
	expect(synth['postcss.config.ts']).toMatchSnapshot()
})
