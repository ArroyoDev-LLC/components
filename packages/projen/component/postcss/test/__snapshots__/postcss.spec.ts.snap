// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders as expected 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

export default {}"
`;

exports[`renders as expected with merged config 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

export default {
    to: \\"something\\",
    from: resolveFrom()
}"
`;

exports[`renders with plugins as expected 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

import postcssPxToRem from \\"postcss-pxtorem\\";
import tailwindcss from \\"tailwindcss\\";

export default {
    plugins: [tailwindcss, postcssPxToRem({\\"propList\\":[\\"font\\"]})]
}"
`;
